lib/jsdom/living/nodes/Document-impl.js

```diff
@@ -140,6 +282,128 @@ class RequestManager {
   }
 }
 
+class PerDocumentResourceLoader {
+  constructor(document) {
+    this._document = document;
+    this._defaultEncoding = document._encoding;
+    this._resourceLoader = document._defaultView ? document._defaultView._resourceLoader : new ResourceLoader();
+    this._requestManager = document._requestManager;
+    this._queue = document._queue;
+    this._deferQueue = document._deferQueue;
+    this._asyncQueue = document._asyncQueue;
+
+    // TODO: Remove when _hasFeature is no longer necessary
+    this._implementation = document._implementation;
+  }
+
+  fetch(url, options) {
+    const { element } = options;
+
+    if (!this._implementation._hasFeature("FetchExternalResources", element.tagName.toLowerCase())) {
+      return null;
+    }
+
+    if (this._implementation._hasFeature("SkipExternalResources", url)) {
+      return null;
+    }
+
+    const requestOptions = {
+      jar: wrapCookieJarForRequest(this._document._cookieJar),
+      proxy: this._document._proxy,
+      strictSSL: typeof this._document._strictSSL === "boolean" ? this._document._strictSSL : true
+    };
+
+    const request = this._resourceLoader.fetch(url, {
+      element,
+      requestOptions,
+      referrer: this._document.URL
+    });
+
+    this._requestManager.add(request);
+
+    const ev = this._document.createEvent("HTMLEvents");
+
+    const onError = error => {
+      this._requestManager.remove(request);
+
+      if (options.onError) {
+        options.onError(error);
+      }
+
+      // TODO: Create events in the modern way https://github.com/jsdom/jsdom/pull/2279#discussion_r199969734
+      ev.initEvent("error", false, false);
+      element.dispatchEvent(ev);
+
+      const err = new Error(`Could not load ${element.localName}: "${url}"`);
+      err.detail = error;
+      err.type = "resource loading";
+
+      this._document._defaultView._virtualConsole.emit("jsdomError", err);
+
+      return Promise.resolve();
+    };
+
+    const onLoad = data => {
+      this._requestManager.remove(request);
+
+      this._addCookies(url, request.response ? request.response.headers : {});
+
+      if (options.onLoad) {
+        try {
+          const result = options.onLoad(data);
```

Comment:
It's a bit sad that the callers have to use onLoad callbacks instead of promises, but this is a good compromise, given that you have to run code after they do.

Reply:
I know, I didn't find a better way to do that :(